### PR TITLE
Enable multi-process training.

### PR DESCRIPTION
**Pull request by archenroot**
On high-count CPU machines stable_baseline supports multi process execution for faster processing.

### CODE REVIEW

**Comment by archenroot** [comment]
Ok, there is now issue with statement:
```
Traceback (most recent call last):
  File "train.py", line 75, in <module>
    while not done:
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

I will investigate, test and do new pull request when fixed

**Comment by archenroot** [comment]
I found the issue, its because done is array of bool now..

**Comment by archenroot** [comment]
Another issue occurred after that fix: 
`TypeError: Cannot cast ufunc add output from dtype('float64') to dtype('int64') with casting rule 'same_kind'`

Testing...

File: train.py [hunk 1]
```diff
@@ -34,11 +34,14 @@
 train_df = df[:train_len]
 test_df = df[train_len:]
 
-train_env = DummyVecEnv([lambda: BitcoinTradingEnv(
-    train_df, reward_func=reward_strategy, forecast_len=int(params['forecast_len']), confidence_interval=params['confidence_interval'])])
+# Enable multiprocess environment
+n_cpu = 32
```
Comment (laneshetron) [inline train.py]: Should probably use `multiprocessing.cpu_count()` here.

File: train.py [hunk 2]
```diff
@@ -64,10 +67,10 @@
     obs = test_env.reset()
     done, reward_sum = False, 0
 
-    while not done:
+    while not np.all(done):
```
Comment (laneshetron) [inline train.py]: Just a heads-up, in my own testing this eventually led to the environments never being marked "done," and evaluation continuing indefinitely - not sure if that's a bug in my code or in Baseline's, though.

Comment (archenroot) [reply train.py]: Ok, I will review ... acually what is happening, that I already have 2 trained agent files, and it still continue, I do for first time this process, I expected there will be just one. How many agents are generated by the train as soon as it gets only best trial hyperparams?

File: train.py (hunk 1 above)
Comment (archenroot) [reply train.py]: or I will provide some variable, if not set (check for None) => multiprocessing.cpu_count()

File: train.py (hunk 2 above)
Comment (archenroot) [reply train.py]: Maybe only one of those in array must be done, the one process which is really last... I need to learn more about the lib

Comment (archenroot) [reply train.py]: It will be good to provide some standard logging into file, for debug as well..

Comment (archenroot) [reply train.py]: I captured output on creation of third file:
```
done type is <class 'numpy.ndarray'>
done is [ True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True]
reward type is <class 'numpy.ndarray'>
reward is [  0.         -60.57482347   0.         -60.57479329   0.
   0.           0.         -66.18156843   0.           0.
   0.         -66.18156846 -66.18156843   0.           0.
 -60.57482328   0.         -60.56543533 -66.18156843 -60.57482347
   0.           0.           0.           0.           0.
   0.         -60.57482347   0.         -60.57296094   0.
   0.         -60.57482347]
[ 2 ] Total reward:  1317274.274415498  (sortino)
[ 3 ] Training for:  13145  time steps

```

Comment (archenroot) [reply train.py]: Interesting is that if I do binary diff, the files are not same:
`cmp -l ppo2_sortino_0.pkl ppo2_sortino_1.pkl | gawk '{printf "%08X %02X %02X\n", $1, strtonum(0$2), strtonum(0$3)}'`

Comment (archenroot) [reply train.py]: And sorry for maybe stupid assumptions, I am playing with this project for 1 day :-))

Comment (laneshetron) [reply train.py]: That's good you're not seeing the issue. I actually wonder if it's my hyperparameter settings, though that doesn't make much sense to me :thinking: . FWIW, definitely try for 4 or 5 episodes, for me it weirdly finishes the first few episodes just fine before running forever beyond that. That said if it works then awesome!

Comment (archenroot) [reply train.py]: Another observation is consequences in **test.py** file:
```
esting PPO2 agent with params: {'cliprange': 0.33306084606428926, 'confidence_interval': 0.8142935150219354, 'ent_coef': 8.523505536828823e-08, 'forecast_len': 3.684454342725584, 'gamma': 0.9619337585646649, 'lam': 0.887423224536798, 'learning_rate': 0.2281478850323017, 'n_steps': 202.464749257154, 'noptepochs': 4.986806900191858}
Best trial: 5380069888.0
Traceback (most recent call last):
  File "test.py", line 49, in <module>
    model = PPO2.load('./agents/ppo2_' + reward_strategy + '_' + str(curr_idx) + '.pkl', env=test_env)
  File "/home/zangetsu/proj/prometheus-core/demo/demo-12-bitcoin-trading-agent/venv/lib/python3.6/site-packages/stable_baselines/common/base_class.py", line 560, in load
    model.set_env(env)
  File "/home/zangetsu/proj/prometheus-core/demo/demo-12-bitcoin-trading-agent/venv/lib/python3.6/site-packages/stable_baselines/common/base_class.py", line 102, in set_env
    "Error: the environment passed must have the same number of environments as the model was trained on." \
AssertionError: Error: the environment passed must have the same number of environments as the model was trained on.This is due to the Lstm policy not being capable of changing the number of environments.

```

So I eventually applied same code as in train.py for multiprocessing, would like to see it how it ends up. ;-)

Comment (archenroot) [reply train.py]: Eventually after make test multi-process as well, it goes well, until hit render:
```
Traceback (most recent call last):
  File "test.py", line 60, in <module>
    test_env.render(mode="human")
  File "/home/zangetsu/proj/prometheus-core/demo/demo-12-bitcoin-trading-agent/venv/lib/python3.6/site-packages/stable_baselines/common/vec_env/subproc_vec_env.py", line 128, in render
    bigimg = tile_images(imgs)
  File "/home/zangetsu/proj/prometheus-core/demo/demo-12-bitcoin-trading-agent/venv/lib/python3.6/site-packages/stable_baselines/common/tile_images.py", line 15, in tile_images
    n_images, height, width, n_channels = img_nhwc.shape
ValueError: not enough values to unpack (expected 4, got 1)
```

Comment (laneshetron) [reply train.py]: Ahh yeah, I've looked into both of those issues a bit. Apparently it's currently a limitation of Stable Baselines that testing requires the same number of environments that a model was trained with (i.e. `n_cpu`).
The second issue was also a bug in Baselines that they ostensibly fixed in the most recent release, but I suspect that it's actually still broken. Not sure :man_shrugging:

**Comment by notadamking** [comment]
@archenroot How's this coming along?

**Comment by archenroot** [comment]
@notadamking - wast testing other stuff the optuna RDBMS with poolsize for shared db where multiple instances (servers) can run same study and and result is centralized, it works, lets see when PR merged.

I am not sure about at moment, lets leave this open for some time, it is definitely which should be addressed. I will at least test without optuna

while not done (arrray of boolean) -> the fact that it continue running even when all are True looks like suspicious...

**Comment by notadamking** [comment]
@archenroot, the alternative route is to use `hyperopt` instead of `optuna`, which already support Postgres db, though I am unsure if we want to do this as `hyperopt` does not take advantage of pruning and can often take much longer to optimize.

**Comment by rshtirmer** [comment]
I've had success training the model using multiple CPUs with these changes, however, I haven't been able to get render() working in test.py, even after changing dummyvecenv to subprocvecenv and looping over num CPUs.

**Comment by archenroot** [comment]
yeah, there is still something to do, I am going to sit on it today...

**Comment by notadamking** [comment]
I am going to close this PR in favor of converging to a single multi-processing PR.

**Pull request closed without merging**